### PR TITLE
Factor out `CompilationUnit` definitions into implementation file

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59465
* __->__ #59245
* #59244

Differential Revision: [D28797510](https://our.internmc.facebook.com/intern/diff/D28797510)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

**Comment by malfet** [comment]
Please add a new file to `core_sources_full_mobile` in `build_variables.bzl`
https://github.com/pytorch/pytorch/blob/8805093ec5c79f689824b6169ddc81bd44067dff/tools/build_variables.bzl#L158

**Review by malfet — APPROVED** [review]

File: torch/csrc/jit/api/compilation_unit.cpp [hunk 1]
```diff
@@ -0,0 +1,373 @@
+#include <torch/csrc/jit/api/compilation_unit.h>
```
Comment (gmagogsfm) [inline torch/csrc/jit/api/compilation_unit.cpp:1]: Thanks for making our implementation more organized!

File: torch/csrc/jit/api/compilation_unit.cpp [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+#include <torch/csrc/jit/frontend/ir_emitter.cpp>
```
Comment (gmagogsfm) [inline torch/csrc/jit/api/compilation_unit.cpp:2]: Curious about the decision to include `cpp` file. Doing so can invite unwanted side effects.

https://www.cplusplus.com/forum/general/39618/

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 3]
```diff
@@ -3360,370 +3360,372 @@ struct to_ir {
   }
 };
 
-CompilationUnit::CompilationUnit(const std::string& source)
-    : CompilationUnit() {
-  // calles the define with native resolver to generate the graph for functions
-  define(c10::nullopt, source, nativeResolver(), nullptr);
-}
+//CompilationUnit::CompilationUnit(const std::string& source)
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3363]: Should we delete code instead of commenting them out?

File: tools/build_variables.bzl [hunk 4]
```diff
@@ -155,7 +155,7 @@ core_sources_full_mobile = [
     "torch/csrc/jit/frontend/convert_to_ssa.cpp",
     "torch/csrc/jit/frontend/exit_transforms.cpp",
     "torch/csrc/jit/frontend/inline_loop_condition.cpp",
-    "torch/csrc/jit/frontend/ir_emitter.cpp",
+    "torch/csrc/jit/api/compilation_unit.cpp",
```
Comment (SplitInfinity) [inline tools/build_variables.bzl:158]: While we're moving code around, what are people's opinions on keeping `CompilationUnit` in the `api` folder? It is part of the public API, but not to the same extent as `Module`, `Object`, and even `Method`.

**Review by SplitInfinity — COMMENTED** [review]
Thanks for doing this. I think it will be ready to land once the instance of `#include` with a `.cpp` is fixed and the commented out code in `ir_emitter.cpp` is removed.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 3 above)
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3363]: Oops, thanks for catching this!

File: torch/csrc/jit/api/compilation_unit.cpp (hunk 2 above)
Comment (ansley) [reply torch/csrc/jit/api/compilation_unit.cpp:2]: It was very difficult to get the headers right, and I realized that everything Just Worked™ when I included the `cpp` file. I didn't know there were such serious consequences. Thanks for the link!

Does this also apply to tests? I have a `cpp` file in `test_cleanup_passes.cpp`. If I use the header file instead, then the implementation isn't linked by the time the test is compiled. If I put the definition in the header, then there's a problem with `RunCleanupPasses` being defined twice

Comment (SplitInfinity) [reply torch/csrc/jit/api/compilation_unit.cpp]: > I have a cpp file in `test_cleanup_passes.cpp`. If I use the header file instead, then the implementation isn't linked by the time the test is compiled.

This suggests that some change needs to be made in some `CMakeLists.txt` to add the new dependency. 🤔

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

File: torch/csrc/jit/api/compilation_unit.cpp (hunk 2 above)
Comment (ansley) [reply torch/csrc/jit/api/compilation_unit.cpp]: Hmm, I'd assume that `run_cleanup_passes.cpp` would need to be specified in `build_variables.bzl::core_sources_full_mobile`, but that doesn't seem to be doing it. 

cc @malfet

File: test/cpp/jit/test_constant_pooling.cpp [hunk 5]
```diff
@@ -93,7 +93,7 @@ graph():
 }
 
 // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-TEST(constantPoolingTest, DictConstantPooling) {
+TEST(constantPoolingTest, DictconstantPooling) {
```
Comment (gmagogsfm) [inline test/cpp/jit/test_constant_pooling.cpp:96]: Accident? It seems DictConstantPooling is correct CamelCase.

**Review by gmagogsfm — APPROVED** [review]
Looks good. Just one minor comment and some tests (due to merge conflicts) need to be fixed.

File: torch/csrc/jit/api/compilation_unit.cpp (hunk 2 above)
Comment (ansley) [reply torch/csrc/jit/api/compilation_unit.cpp]: Resolved by adding `TORCH_API`

File: test/cpp/jit/test_constant_pooling.cpp (hunk 5 above)
Comment (ansley) [reply test/cpp/jit/test_constant_pooling.cpp:96]: Good catch!

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797510).

**Pull request closed without merging**